### ISSUE TITLE
Re-upstreaming of the KGPE-D16

### ISSUE THREAD

**Issue by tlaurion** [osresearch/heads#719]
There was a proposition to do required work under refused grant application. If there is interest from third party to fund conjointly needed work, board could be maintained and reupstreamed, both under u-bmc and coreboot.

Putting it out there.

**Comment by tlaurion** [comment]
Traces:
[u-bmc](https://github.com/u-root/u-bmc/issues/133)
[kgpe-d16 actual board status](https://github.com/osresearch/heads/issues/134)

This board works as a QubesOS server while required work is needed.
Freshly landed [QubesOs-network-server](https://github.com/Rudd-O/qubes-network-server/issues) code

A followup on https://github.com/osresearch/heads/issues/717#issuecomment-625825019

**Comment by Tonux599** [comment]
@tlaurion did 3mdeb decide not to take it on? It would be good to have as it's the only completely blob free board I can think of that's qubes compatible. Provided a 6200 series is used.

**Comment by tlaurion** [comment]
@Tonux599 3mbeb is absolutely willing to do the work in exchange of paid work.

**Comment by tlaurion** [comment]
@Tonux599 

NlNet decided to not take it on (Fund the project).
Q: Will we decide to take it on and make it happen? Let's see. :)

**Comment by tlaurion** [comment]
[Applicable](https://ngiatlantic.eu/)?

**Comment by tlaurion** [comment]
@pietrushnic

**Comment by pietrushnic** [comment]
@tlaurion sorry, for not replying, recently we have quite a lot on my plate. Definitely, we would like to publish here documents that we sent to NLNet while applying for a grant. There is also a slightly modified version that we provided to Insurgo on Slack. Both documents rely on coreboot state in December 2019. @miczyg I think you can push grant content to 3mdeb GitHub repo as PR and link here for review.

**Comment by miczyg1** [comment]
This Pull Request contains the application 3mdeb sent to NLNet to obtain the grant:
https://github.com/3mdeb/kgpe-osf/pull/1
Feel free to comment and review.

**Comment by Tonux599** [comment]
@pietrushnic @miczyg1 When should we expect an outcome? Also could you briefly outline any ways that members of the community could support you if or if not you get funding?

**Comment by pietrushnic** [comment]
@Tonux599 there is no timeline since there is no official commitment from anyone to fully found the effort. We get some good words from various individuals and organizations, but that's all. @tlaurion mentioned that he can sponsor part of the effort, but we are not sure anyone would be happy with half-finished stuff. Vikings Gmbh sent us hardware so we have something to work on. NLNet did not approve of this grant, so I would not expect it will change in the future - I may be the wrong here. To sum the status of this project is "not founded" so we even do not put that on schedule and worry more about founded effort e.g. TrenchBoot and fwupd for Qubes OS.

How the community can help? We are very open here, whatever works for the community.
1. We can split this project and drive as an open and free contribution if there are developers and testers who willing to support the development and debugging. In such a situation we will support whoever will do that as much we can including, coordination review and help in merging code upstream.
2. We can open any channel that the community is comfortable with to try to gather remaining founds. @tlaurion suggested OpenCollective, we can align probably to any suggestion - we have no problem in handling the administrative part.
3. We have a PayPal donation button on our website, although this is probably one of the worst options. 
4. We are fine with commercial assignments if anyone is willing to go that path. It can be even a very small part of the task, but at least it will move us forward.
5. If we can sell a platform in some form or bundle services, which will generate margin, then we can use it to bring back the platform to the mainline.

We would like to help, but without funding, we will do the best effort and it may take very long to bring back this platform. Please note our goal is not just to bring back the platform to mainline, but create a structure in which long term maintenance would not cost a lot. We did that for [PC Engines routers](https://pcengines.github.io/) for over 4 years.

**Comment by tlaurion** [comment]
@Tonux599 : For the moment, step would be to get people interested in the port. Let's remember that the KGPE-D16 was funded by [Libreboot and put upstream by RaptorEngineering](https://libreboot.org/docs/hardware/kgpe-d16.html) but previous work did not receive enough attention to be maintained and got dropped in coreboot 4.11 release for lack of maintainership.

Let's also remember that the community crowdfunder the port of the[ BMC chip to OpenBMC ](https://www.raptorengineering.com/coreboot/kgpe-d16-bmc-port-offer.php)and that work wasn't maintained.

NlNet refused to fund the work because they didn't understand the reasoning of funding old platform.

So the point here would be first to get enough attention so that the story doesn't repeat itself.
Insurgo is interested to fund a big part of the work, but we concluded that we would not do this alone.

Putting a bounty tag for this.

**Comment by tlaurion** [comment]
Something really interesting was proposed on this QubesOS ticket: 
https://github.com/QubesOS/qubes-issues/issues/2809#issuecomment-626111758

@pietrushnic, I would love to hear on your reception.

**Comment by pietrushnic** [comment]
@tlaurion XVilka suggestion looks interesting, but definitely key problem is orchestration. We would like to avoid the situation when 3 or more teams work on a given feature in parallel and then post results claiming the founds. There should be serious commitment based on proven reputation and trust, otherwise, we have public bidding problem where lower price wins with quality.

Whonix stuff is even better, but we would have to research how it looks from the taxation and administration perspective.

**Comment by tlaurion** [comment]
@pietrushnic I'm not sure there is conflict between the two ideas with XVilka avenue. 
If the ticket is clear saying that 3mdeb realizes the work in question (milestone) and funds are released upon proof of work (PR). But you are better placed then I am to see if this would work or not.

**Comment by mikebdp2** [comment]
@tlaurion and @pietrushnic : Have you contacted a Free Software Foundation? Since they have a lot of interest in RYF devices and are probably using these KGPE-D16 servers by themselves, if they'd learn about this critical situation - maybe they'll decide to co-fund?

**Comment by pietrushnic** [comment]
@mikebdp2 we (as 3mdeb) never contacted FSF. I'm not sure if the situation is critical but it will get worst over time. From experience longer platform lays unmaintained bigger would be effort to bring it back. Do you know anyone who we should reach?

@tlaurion I believe you are in a better position to talk with FSF, then 3mdeb.

**Comment by pkubaj** [comment]
22000E isn't that much different from what Raptor wanted for OpenBMC port (at https://www.raptorengineering.com/coreboot/kgpe-d16-bmc-port-offer.php). Maybe you should set up similar crowdfounding? Currently you only say "we are ready to work" etc. Raptor did the same and received no money, but when they officially started crowdfounding, it turned out there were enough people interested in it to gather the money.

**Comment by pietrushnic** [comment]
@pkubaj very good point. It looks like Martin was intermediary gating the process. In case of difference in pricing, please note this was a different time and they are in the US not in PL, their costs are way higher than ours. More to that we are Yocto Participants and have Embedded Linux people on board, OpenBMC uses Yocto/OE build system, so this also simplifies the situation for 3mdeb.

**Comment by tlaurion** [comment]
> @mikebdp2 we (as 3mdeb) never contacted FSF. I'm not sure if the situation is critical but it will get worst over time. From experience longer platform lays unmaintained bigger would be effort to bring it back. Do you know anyone who we should reach?
> 
> @tlaurion I believe you are in a better position to talk with FSF, then 3mdeb.

@pietrushnic writing to them.

**Comment by alexmaloteaux** [comment]
> @Tonux599 : For the moment, step would be to get people interested in the port. Let's remember that the KGPE-D16:
> 
>     * was funded by [Libreboot and put upstream by RaptorEngineering](https://libreboot.org/docs/hardware/kgpe-d16.html) but previous work did not receive enough attention to be maintained and got dropped in coreboot 4.11 release for lack of maintainership.
> 
>     * the community crowdfunded the port of the[ BMC chip to OpenBMC ](https://www.raptorengineering.com/coreboot/kgpe-d16-bmc-port-offer.php)and that work wasn't maintained.
> 
>     * NlNet refused to fund the work because they didn't understand the reasoning of funding old platform.
> 
> 
> So the point here would be first to get enough attention so that the story doesn't repeat itself.
> Insurgo is interested to fund a big part of the work, but we concluded that we would not do this alone.
> 
> Some history:
> 
>     * [FOSDEM talk by 3mdeb on AMD coreboot status](https://fosdem.org/2020/schedule/event/coreboot_amd/)
> 
>     * [Reddit thread on announced drop of the KGPE-d16](https://www.reddit.com/r/linux/comments/dz2hlt/boards_like_asus_kgped16_the_most_powerful/)
> 
> 
> Putting a bounty tag for this.

Hi is there a list of the bugs/linux crash that Michal talk about in his fosdem talk ? Do they only happen when iommu is enabled ?

Btw i have a kgpe-d16 with jtag header soldered and an olimex. So if you are starting any group  for dev , put me in, free time is limited but i can always help a bit and test.

**Comment by tlaurion** [comment]
@miczyg1 ?

**Comment by tlaurion** [comment]
@alexmaloteaux asked:
> Hi is there a list of the bugs/linux crash that Michal talk about in his fosdem talk ? Do they only happen when iommu is enabled ?
@miczyg1 ?

**Comment by miczyg1** [comment]
Simply have a look at the coreboot mailing list:
- https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/ME4TULTDZGIYJEMNR2IB2W6P7ONG76EU/#GCJF7CIWGCKZCDWYPWAM2LKQW6APHZRJ (memory problems)
- https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/ZR7LMBXOFKR5FWVQDI22UB6O3LO5WGQ3/#TH57VEGCY357ZH534PU5PKSDSMYLTLFH (S3 suspend/resume problems)
- https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/VWFKH7FDWLO5KGZMIJ7VM67SIUWJRVZL/#OCHSI6UL2XZDSA3HCHOONNN5EX4HLUGE (kernel panic from new microcode)
- https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/H5KHOUVDPQA2NSUE43IBRKSYBZQK4DWN/#H5KHOUVDPQA2NSUE43IBRKSYBZQK4DWN (SLUB memory allocator in Linux panics)
- https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/RDHEDOTBYVD74NWQWYYNLAJWPD7CMDRZ/#RDHEDOTBYVD74NWQWYYNLAJWPD7CMDRZ (KVM/SMV not working)
- https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/AW4DF557YDTTJGKDFRTHMZJBTF65ZJZT/#VOFCIRV24HT3OIG5IDYH4GEXY5VSS3W5 (kernel panics on kernel newer than 4.9)

**Comment by pietrushnic** [comment]
In case of what have to be done I believe Arthur email is quite good. @tlaurion you replied to it [here](https://www.mail-archive.com/coreboot@coreboot.org/msg54125.html)

**Comment by alexmaloteaux** [comment]
I have read the various mail list entry and i cant reproduce any of them. Actually i have a pretty stable system so far so i will list here my config and test , maybe others can use it as baseline to try to triggers issues and/or star the long 4.12 migration process. please note i have a pure server usage for this board and not a desktop pc.

- Coreboot version : 4.10, commit ea58adddf4  with seabios and sgabios, all input/output redirected to uart 0x2f8 io port for OpenBmc (coreboot, sgabios, os)
- Heads Version : master but cant use atm as im facing TPM issue
- CPU : 2X Opteron 6378 2.4GHz 16M
- Ram : 2x Micron 16GB Ram DDR3 MT36KSF2G72PZ-1G6E1FE PC3L-12800R
- TPM : Asus TPM-L r2.0 TPM 20-1 pin (recognized by seabios but tpm-reset script or tpm-tool have issue ; unknown tag ...)
- Screen : Not using VGA at all, all input/output redirect to OpenBmc uart on io port 0x2f8
- Linux Boot : OK (ubuntu 18.04 / kernel 4.15)
- Freebsd Boot : OK (12.1 stable)
- Linux Boot with IOMMU enabled : OK (iommu=pt amd_iommu=on)
- Freebsd Boot with IOMMU : OK
- Linux install and boot VM in KVM : OK
- Freebsd install and boot VM in bhyve : not yet tested
- Linux PCI-passthrough to VM in KVM : OK (tested with a IBM M1015 flashed in it mode)
- Freebsd PCI-passthrough to VM in Bhyve : not yet tested


Ill try to post a more detail description of the tpm issue i m facing later on

BR

**Comment by Tonux599** [comment]
@alexmaloteaux regarding TPM, TPM2 is currently unsupported with heads, see #287 however the Asus 90-C1B0AU-00XBN0VZ is a TPM1.2 module compatible with heads and this board.

**Comment by pkubaj** [comment]
I have two of those boards, one is a desktop, the other one is a server. Both run coreboot 4.11 tag.

Both run FreeBSD, the desktop is sometimes booted to Linux.

Both are stable. I didn't test averything that @alexmaloteaux tested. I'm pretty sure PCI passthrough won't work with bhyve due to CPU's missing some required features.

**Comment by alexmaloteaux** [comment]
> I have two of those boards, one is a desktop, the other one is a server. Both run coreboot 4.11 tag.
> 
> Both run FreeBSD, the desktop is sometimes booted to Linux.
> 
> Both are stable. I didn't test averything that @alexmaloteaux tested. I'm pretty sure PCI passthrough won't work with bhyve due to CPU's missing some required features.

63xx Series supports  AMD-Vi and can support PCI passthrough , except os/coreboot bug

**Comment by Tonux599** [comment]
> > @mikebdp2 we (as 3mdeb) never contacted FSF. I'm not sure if the situation is critical but it will get worst over time. From experience longer platform lays unmaintained bigger would be effort to bring it back. Do you know anyone who we should reach?
> > @tlaurion I believe you are in a better position to talk with FSF, then 3mdeb.
> 
> @pietrushnic writing to them.

@tlaurion any luck?

**Comment by tlaurion** [comment]
Waiting for an e-mail reply from FSF

**Comment by tlaurion** [comment]
I would suggest anyone interested in the d16 to be reupstreamed writing to FSF too?

**Comment by Tonux599** [comment]
> I would suggest anyone interested in the d16 to be reupstreamed writing to FSF too?

@tlaurion any chance of seeing what you wrote or if not guidelines on what would good to include in the correspondence?

**Comment by tlaurion** [comment]
> > I would suggest anyone interested in the d16 to be reupstreamed writing to FSF too?
> 
> @tlaurion any chance of seeing what you wrote or if not guidelines on what would good to include in the correspondence?

@Tonux599: 
Contacted them through patron@fsf.org and business@fsf.org.

ryf@fsf.org might be of interest.

```
Hello FSF,


Insurgo here!


We have tried to get NlNet fund the KGPE-D16 revival. As you may know,
coreboot dropped support for that board in coreboot 4.11, leaving the
last RYF platform unmaintained and now rotting.

3mdeb made a really good grant application, but NlNet didn't really
understood the importance of maintaining old hardware, and unfortunately
the ball got dropped.


The details are in this ticket:
https://github.com/osresearch/heads/issues/719


I would love to start a discussion with you guys on what is possible and
if funds were available from you to jump in this subproject.


Thierry

Insurgo Open Technologies
```

**Comment by alexmaloteaux** [comment]
How can we write them ? everybody individually pointing at this issue ? or  collectively ?

Best Regards

**Comment by tlaurion** [comment]
@alexmaloteaux from my understanding of the [past](https://github.com/osresearch/heads/issues/719#issuecomment-627605335), interest from endusers and company should be expressed individually by the most numerous number of people and why it matters.

**Comment by tlaurion** [comment]
@pietrushnic @miczyg1 : took some time to give love to #472 in link of #712. We are getting near of a heads baseline.

**Comment by tlaurion** [comment]
RYF contacted and put 3mdeb in contact directly. Let's see where it goes...

**Comment by tlaurion** [comment]
@pietrushnic @miczyg1 there is a call preparation with RMS from FSF, vikings.net and tiberiu from tehnoetic.com

I advised that you join the discussion and shared Piotr e-mail address. Will follow through.

**Comment by ullbeking** [comment]
@alexmaloteaux 
> Btw i have a kgpe-d16 with jtag header soldered

Alexandre, I was really impressed to read that you got the JTAG header installed on the ASUS KGPE-D16 board :-)  I assume that you also had to remove the solder plugs from the board first?  If you are reading this message, and if you are willing to give a little advice, please could you send me a sign?!  :-)  I have been preparing the board to remove the solder plugs safely, and I would love to compare notes with somebody who has already done it successfully.

**Comment by Th3Fanbus** [comment]
@tlaurion about the e-mail to FSF:

> We have tried to get NlNet fund the KGPE-D16 revival. As you may know,
coreboot dropped support for that board in coreboot 4.11, leaving the
last RYF platform unmaintained and now rotting.

The chosen wording makes it seem as if it is the coreboot project's fault that the KGPE-D16 ended up unmaintaned and rotting. However, I have seen no one maintaining KGPE-D16 support in upstream coreboot, only users complaining about things not working. That the code for this mainboard got dropped from coreboot after the 4.11 release is not the reason why the board is no longer maintaned, but the other way around: as there had been no maintenance going on for a rather long time (even though there were plenty of bugs to fix), it had become a maintenance burden for coreboot developers and the code eventually got dropped as a result. Quite surprising, given that several companies are making profit by putting coreboot onto these boards.

Moreover, it is not the last RYF platform. It might be the most powerful x86 platform, but definitely not the only one.

**Comment by pkubaj** [comment]
> @tlaurion about the e-mail to FSF:
> 
> > We have tried to get NlNet fund the KGPE-D16 revival. As you may know,
> > coreboot dropped support for that board in coreboot 4.11, leaving the
> > last RYF platform unmaintained and now rotting.
> 
> The chosen wording makes it seem as if it is the coreboot project's fault that the KGPE-D16 ended up unmaintaned and rotting. However, I have seen no one maintaining KGPE-D16 support in upstream coreboot, only users complaining about things not working. That the code for this mainboard got dropped from coreboot after the 4.11 release is not the reason why the board is no longer maintaned, but the other way around: as there had been no maintenance going on for a rather long time (even though there were plenty of bugs to fix), it had become a maintenance burden for coreboot developers and the code eventually got dropped as a result. Quite surprising, given that several companies are making profit by putting coreboot onto these boards.
> 
> Moreover, it is not the last RYF platform. It might be the most powerful x86 platform, but definitely not the only one.

What were those bugs? I have coreboot 4.11 working on two KGPE-D16 boards. It has some issues, but they are easy to workaround. Those boards are quite finicky to set up with RAM compatibility issues etc., but this is also the case with vendor firmware.

I think those "reports" about broken coreboot were just users doing things wrong. And coreboot developers, instead of noticing that it works for other users (I repeatedly reported that) decided to go the easy way - just nuke it.

There's also another thing. I'm not a firmware developer, so I wouldn't be able to fix any issues in coreboot. But it happens that I'm a FreeBSD ports developer and I often fix ports that I don't care about, but I know users care. coreboot developers didn't have that sort of attitude. I wonder why. I just hope that there wasn't anyone sabotaging coreboot, because I would be sad to see that.

**Comment by paulmenzel** [comment]
I vote for closing this issue, as too much FUD is spread.

> > @tlaurion about the e-mail to FSF:
> > > We have tried to get NlNet fund the KGPE-D16 revival. As you may know,
> > > coreboot dropped support for that board in coreboot 4.11, leaving the
> > > last RYF platform unmaintained and now rotting.
>
> > The chosen wording makes it seem as if it is the coreboot project's fault that the KGPE-D16 ended up unmaintaned and rotting.

I agree that the wording is not optimal.

[…]

> What were those bugs? I have coreboot 4.11 working on two KGPE-D16 boards. It has some issues, but they are easy to workaround. Those boards are quite finicky to set up with RAM compatibility issues etc., but this is also the case with vendor firmware.
> 
> I think those "reports" about broken coreboot were just users doing things wrong. And coreboot developers, instead of noticing that it works for other users (I repeatedly reported that) decided to go the easy way - just nuke it.

We talked about this in #coreboot@irc.freenode.net quite often: Random boot failures, all of memory not supported, higher energy usage compared to vendor firmware, problems loading latest microcode, ….

> There's also another thing. I'm not a firmware developer, so I wouldn't be able to fix any issues in coreboot. But it happens that I'm a FreeBSD ports developer and I often fix ports that I don't care about, but I know users care. coreboot developers didn't have that sort of attitude. I wonder why. I just hope that there wasn't anyone sabotaging coreboot, because I would be sad to see that.

Hardware development is much harder than software development. You need to test the changes on actual hardware. Chipset changes affect other boards too, so you have to test those too. Also you need to have understanding of the hardware.

You can get active, and push the collecting the money forward by starting some kind of crowdfunding.

**Comment by pkubaj** [comment]
> We talked about this in #coreboot@irc.freenode.net quite often: Random boot failures, all of memory not supported, higher energy usage compared to vendor firmware, problems loading latest microcode, ….

Random boot failures - that happens to me maybe in 1% or so of boot attempts (I mean, very rarely). True, vendor firmware doesn't do that, but this is so low priority to me that I literally don't care,
All of memory not supported - isn't that also the case with vendor firmware? That would mean there's a hardware issue, which coreboot can at best workaround. Still there ARE RAM sticks that work properly,
higher energy usage - that's unfortunately true, but that causes energy bills to go up, it doesn't cause any usage issues,
Problems loading latest microcode - I don't have this issue, I have coreboot built with the latest microcode and it works.

Sure, there ARE issues that prevent coreboot from achieving 100% feature parity with vendor firmware. But, this is also the case with other boards. On ASUS F2A85-M I had to manually run fancontrol on Linux because CPU would overheat otherwise and FreeBSD was unusable on it (due to no fancontrol). Yet, I can see F2A85-M is still in upstream coreboot.

On ThinkPad X230, AFAIK the yellow USB port (that can charge devices even when the laptop is off) works like all other USB ports (so it charges only when the laptop is on). X230 is also still in upstream coreboot.

So IMO arguments that there are bugs don't hold as long as it's actually possible to get the board working and users are reporting them working.

**Comment by paulmenzel** [comment]
Please let’s talk in IRC, before this issue gets even more convoluted.

You totally misunderstand, why the board was removed from the main branch (but it’s still in the branch `4.11_branch`).

It was *not* about the bugs or an incomplete port, it’s because the developers decided that each platform/board has to support certain features. Please read the [release notes](https://review.coreboot.org/cgit/coreboot.git/tree/Documentation/releases/coreboot-4.12-relnotes.md#n72).

> > We talked about this in #[coreboot@irc.freenode.net](mailto:coreboot@irc.freenode.net) quite often: Random boot failures, all of memory not supported, higher energy usage compared to vendor firmware, problems loading latest microcode, ….
> 
> Random boot failures - that happens to me maybe in 1% or so of boot attempts (I mean, very rarely). True, vendor firmware doesn't do that, but this is so low priority to me that I literally don't care,

You asked, what bugs, and claimed there are none. Come on. That you live with them, is another topic.

> All of memory not supported - isn't that also the case with vendor firmware? That would mean there's a hardware issue, which coreboot can at best workaround. Still there ARE RAM sticks that work properly,

The vendor firmware supports all 256 GB of memory. Where did you get this misinformation from? And again, how is your last statement even related?

> higher energy usage - that's unfortunately true, but that causes energy bills to go up, it doesn't cause any usage issues,
> Problems loading latest microcode - I don't have this issue, I have coreboot built with the latest microcode and it works.

Good for you. I guess you do not use the Linux kernel?

> Sure, there ARE issues that prevent coreboot from achieving 100% feature parity with vendor firmware. But, this is also the case with other boards. On ASUS F2A85-M I had to manually run fancontrol on Linux because CPU would overheat otherwise and FreeBSD was unusable on it (due to no fancontrol). Yet, I can see F2A85-M is still in upstream coreboot.
> 
> On ThinkPad X230, AFAIK the yellow USB port (that can charge devices even when the laptop is off) works like all other USB ports (so it charges only when the laptop is on). X230 is also still in upstream coreboot.
> 
> So IMO arguments that there are bugs don't hold as long as it's actually possible to get the board working and users are reporting them working.

See my statement at the very top. No idea, where you got your information about the cause for removal from. And people can still use the version from 4.11 (even up to the removal commit). Why do they need the new features, if it works as you said?

Thanks to the payload concept, people can still use newer payloads, which makes the main difference to them.

Anyway, I am out, and kindly ask you to inform yourself more diligently about the whole situation.

**Comment by ullbeking** [comment]
@paulmenzel 
> Please let’s talk in IRC, before this issue gets even more convoluted.

If we are going to move the discussion then the mailing list would be vastly better.  It is important to correct the misunderstandings on the record.  IRC is too "whizzy" to be able to absorb the complexities of the problems and the social interactions that led to this problem in the first place.

It's a convoluted issue and I think participants at all levels of experience are still trying to understand what's really going on.  If this had been discussed on a mailing list then I would at least have been aware of the problems.  If on IRC then no chance, because I'm not on IRC nearly as much as I used to be (I just **can't**).

_Personally, I am trying to get much more actively involved in this, especially since I recently realized that the lack of support for the D16 port of coreboot was much less supported than I originally thought.  **It was only by reading the complicated conversation -- which is necessary to tease out the misunderstandings -- that I realized the nature of the problem AT ALL.**_

I don't think that shifting the conversation to IRC is going to help clarify the issue **for me.**  And I REALLY want to get involved with this.  In fact, I think it will make it worse because as my daughter is about to start school and family life is becoming important I am unable to spend sll day every day on IRC.  (Unlike the last several years.)

**Comment by pietrushnic** [comment]
@Th3Fanbus 

>  Quite surprising, given that several companies are making profit by putting coreboot onto these boards.

It is easy to write that, but hard to prove `profit` hypothesis. Hardware margins are not so big and I'm pretty sure the volume of those boards does not justify further investment especially when the previous one failed.

@ullbeking 

> In summary, the complexity of this conversation and working through it is actually necessary to understand the nature of the problem, which is just as much a social issue as a technical one. 

I agree with social part, technical is little bit easier for us (3mdeb). There is too much, not needed, misunderstanding and throwing rocks. IMO the community is about trust, responsibility, and eventually reputation. I'm sure we all are acting in goodwill here, but there should be an understanding that no company can afford infinite firmware support without a net profit margin. We believe it is possible to create after-market support driven by the community if it is big enough. That approach is hard like any B2C project.

In an ideal world, we would have enough traction in B2B relations to support the community. There are success stories behind that e.g. [PC Engines](https://pcengines.github.io/), which we support for more than 4 years. There is [Olimex and Allwinner story presented at FOSDEM 2020](https://fosdem.org/2020/schedule/event/olimex_oshw/) that IMO applies here. To drive this success we need people like @tlaurion, coordinators, and organizers not in a chaotic way but well-organized project/product management way. Otherwise, we just have chaos or count on luck that NLNet or FSF will support.

Someone may ask if 3mdeb knows that why it will not organize everything? The answer is that we already delivered as much as we can and we plan to keep our promises to some point, despite time passing by and situation is more and more complex. We also have limited resources, which we have to manage wisely to deliver to the community great new open-source firmware projects/products.

What we can do? We should look at our peers, especially business peers, and figure out how they can benefit from well-maintained KGPE-D16, or maybe you have an idea for a business (even very small one) based on fully open platform. Of course, the goal is to get even small monetary contributions from them and gather the whole amount to reenable KGPE-D16 in upstream coreboot. We open to talk with those peers and negotiate conditions, then inform community about the progress.

We are not experienced with crowdfunding, but eventually will get there, if someone knows that environment we would be glad to talk about that. At some point, @tlaurion pointed to [this campaign](https://www.raptorengineering.com/coreboot/kgpe-d16-bmc-port-offer.php) which in my opinion is a great example of how things can be organized.

Long run (12 months +) we can have a bug bounty program that will gather founds for maintainership and bug fixing.

**Comment by Tonux599** [comment]
In relation to the point above, @tlaurion has Insurgo considered a PrivacyBeast KGPE-D16?

**Comment by alexmaloteaux** [comment]
Hi sorry for asking but until there is a final place to ask question , can somebody tell me if there is any solution/workaround for the energy usage.

im using cb 4.10 and i lowered the frequency of my 2x6378 to 1400 with powerd on hardenedbsd and it is mostly ok . But at 2400 Hz i need to have my dynatron g34 heatsink running at 100% all the time and the small room where the server is located is becoming really hot after a few hours of usage.

Thanks

**Comment by tlaurion** [comment]
Of course,  but the actual situation of unknown stable sourcing and required work to make the kgpe-d16 maintained and back into coreboot is out of my personal scope.

Coreboot in heads is now.pinned to specific version in master. So the work to make the kgpe-d16 work with coreboot 4.11 can then start by the community, on which I will help, but won't pioneer without funding. 

If someone here wants to help, then it goes back to starting a crowdfunding page, on which Insurgo would jump in and fund part of the work too.

Unfortunately, this is the situation we are in right now. My focus is more on current blob free laptops and future proof platforms. But would gladly work conjointly on bringing back to life the kgpe-d16.

Still no news from the FSF as of now and unfortunately went the farthest I could from a community, unpaid, perspective on my own for this platform. 

Any help/tangible plan welcome on making this funded partially / conjointly would be awesome. Otherwise im a bit overwhelmed by other tasks to lead on this one.

Thierry/Insurgo

On July 15, 2020 1:11:58 PM UTC, Tonux599 <notifications@github.com> wrote:
>In relation to the point above, @tlaurion has Insurgo considered a PrivacyBeast KGPE-D16?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/osresearch/heads/issues/719#issuecomment-658758622
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by agg1** [comment]
Not that anyone asked but I feel urged to drop a note concerning support for various AMD AM3/G34 boards. AM4 generation and similar are dead already anyway as far as i am concerned. same holds true for anything equipped with ME and similar.
1) branch 4.11 exists and can be contributed to and i welcome the fact it could be considered "stable" now for final review
2) alltogether the quality of what's present for AMD AM3/G34 after a decade of development raises concerns by itself, given this branch reached deprecated state with a quality present which I had considered not ready for production when this gear reached markets 10 years ago. the only difference to vendor bios i see concerning this is vendors do not provide sources. and to prevent a misunderstanding i might add it is certainly not coreboot devs to blame for this situation.
3) some patches for branch 4.11 are worked on to fix some AM3 board (https://github.com/agg1/coreboot/commits/master), which may be a decent alternative to KGPE-D16 and a viable option for a devs who couldn't afford a corebooted workstation for thousands of bucks
4) most importantly i would urge the industry including the FOSS people to rethink their general attitude towards resource requirements, as if it was necessary for amateur hackers and downstream developers to suck kilowatts from the powerwall with a 32core workstation and 256GiB of RAM, meanwhile a Hexacore with 16GiB RAM should do for another decade to come for some system integration tasks and distro maintenance. i can and do maintain a self-made hardened livedvd linux distro and it finished compiling everything from source in less than 24hours and 150Watts power draw only on a corebooted am3 workstation due to some optimizations und cuttings
5) my budget, and even if i could afford a KGPE-D16 i wouldn't want it but instead industry including FOSS people should align their resource requirements so development can be done on commodity hardware such as AM3 boards or quadcore laptops with 16GiGs of RAM with acceptable electricity costs for 24/7 full load operation, which makes me wonder again why up until recently a dual core with 2GB RAM was the only option officially tested with AM3-generation commodity hardware. same holds true for the chaotic nature of documentation available (BIOS, ACPI, chipsets, UEFI) which requires almost a livetime of work to attach some peripherals, initialize memory, etc. it is a nightmare.
6) i could rant day and night about industry habits (20 million LoC for GCC alone). again, rethink if industry could afford to loose another generation of amateur hackers them instead playing video games on threadripper workstations with PSP then

**Comment by tlaurion** [comment]
> @tlaurion about the e-mail to FSF:
> 
> > We have tried to get NlNet fund the KGPE-D16 revival. As you may know,
> > coreboot dropped support for that board in coreboot 4.11, leaving the
> > last RYF platform unmaintained and now rotting.
> 
> The chosen wording makes it seem as if it is the coreboot project's fault that the KGPE-D16 ended up unmaintaned and rotting. 

I am really sorry for the impact those words had. My intention was more to say exactly like everyone here is saying: the KGPE-D16 got dropped because of a lack of consistent maintainsership/testing and other problems IMHO, where others disagree. I think that most of the KGPE-D16 users came from libreboot, which was using a static version of coreboot, which again in my opinion, gave the illusion that those two projects were disjoint. The goal here is to learn from the past, not to blame anybody. When I say that the KGPE-D16 was left rotting, its historicism and I should not do that logic error. The facts are there though, in coreboot port not having enough working status reports in time for it to be maintained, down to crowdfunded OpenBMC port which stayed as a PoC with issues reported, confirmed, fixes provided while not merged for OpenBMC build system; no maintainership was made in time, nor regression testing to prevent dropping. 

This leaves us where we are now, where OpenBMC port was left behind and branched, coreboot port is diverging from upstream from 4.11 (with missing security fixes and incrising costs of putting that board upstream).

It is really a sad story but there needs to be demand and community support at some point.
- [ ] https://github.com/u-root/u-bmc/issues/133

 
> However, I have seen no one maintaining KGPE-D16 support in upstream coreboot, only users complaining about things not working. That the code for this mainboard got dropped from coreboot after the 4.11 release is not the reason why the board is no longer maintained, but the other way around: as there had been no maintenance going on for a rather long time (even though there were plenty of bugs to fix), it had become a maintenance burden for coreboot developers and the code eventually got dropped as a result. Quite surprising, given that several companies are making profit by putting coreboot onto these boards.


> 
> Moreover, it is not the last RYF platform. It might be the most powerful x86 platform, but definitely not the only one.

- There is [3 RYF mainboards](https://ryf.fsf.org/categories/mainboards). 
- Old laptops not supporting virtualization extensions but not having ME per libreboot work.
- coreboot blob-free boards like the [G505S are supported but limited in SPI flash and without a TPM](https://github.com/osresearch/heads/issues/453)... 

Other boards were documented in [earlier posts of this thread](https://github.com/osresearch/heads/issues/719#issuecomment-659024471). 

@miczyg1 made an awesome talk [here](https://fosdem.org/2020/schedule/event/coreboot_amd/) including status of KGPE-D16.
Please read my slides 45+ [here](https://fosdem.org/2020/schedule/event/firmware_hodorateatria/) for a global comprehension of the actual status of blobs in newer boards. RYF on Intel/AMD will not happen for newer boards. I think we should revive the KGPE-D16 and KCMA-D8 as a community but I fought enough on my own and no door opened following my personal attempts to date. I would gladly support someone taking that leadership and moving things forward both with my time and financially but won't lead this myself.

**Comment by lss4** [comment]
Back then I posted some of my efforts trying to get KGPE-D16 working on the mailing list. Unfortunately it did not get me anywhere further than that. It's possible that libreboot still works okay for this board (given its codebase), but I'm not as familiar with libreboot's build environment as with coreboot's.

With coreboot 4.11 (4.11_branch), RAM initialization issues on KGPE-D16/KCMA-D8 are hit-or-miss, at least I can get some 100%-bootable RAM combinations on KGPE-D16 for the time being. However, it seems the real issue isn't RAM and lies in somewhere else. For example, something's not right in libpayload with this board as after looking into the code the fatal error messages I'm getting in some places (like trying to run nvramcui/coreinfo) originated from there (mostly USB related). I'm not an expert in this field, but libpayload might be something worth looking into, for finding out what might have caused coreboot to not work as well as before on this board.

For now I can embed and boot some floppy images using coreboot, thanks to Mike Banon's patches, They appear to be stable, just that USB doesn't work. Other than that there's little I could achieve as I need to boot from USB to install an OS and I haven't gotten myself to successfully boot anything from USB yet (either freeze or with error messages, probably because of the aforementioned libpayload issues).

> some patches for branch 4.11 are worked on to fix some AM3 board (https://github.com/agg1/coreboot/commits/master), which may be a decent alternative to KGPE-D16 and a viable option for a devs who couldn't afford a corebooted workstation for thousands of bucks

> my budget, and even if i could afford a KGPE-D16 i wouldn't want it but instead industry including FOSS people should align their resource requirements so development can be done on commodity hardware such as AM3 boards or quadcore laptops with 16GiGs of RAM with acceptable electricity costs for 24/7 full load operation, which makes me wonder again why up until recently a dual core with 2GB RAM was the only option officially tested with AM3-generation commodity hardware. same holds true for the chaotic nature of documentation available (BIOS, ACPI, chipsets, UEFI) which requires almost a livetime of work to attach some peripherals, initialize memory, etc. it is a nightmare.

One major issue with these workstation/server AMD boards would be the availability of cooling solutions. Nowadays C32/G34 cooling solutions are very hard to find (although someone managed to get some others installed by hard-modding the clips). In comparison, cooling solutions for Intel sockets (115x, 1366, 2011, etc.) and mainstream AMD sockets (AM2/AM3, AM4, FM1, etc.) are much more common. It'd certainly be better if mainstream AMD motherboards of that era could be better supported, I'm looking forward to it.

**Comment by agg1** [comment]
Total power consumption with the corebooted AM3 board with a Phenom 2 X6 of mine is far below any of the server/workstation boards: ~140Watts total power consumption under full load (monitor, 4 hard disks in RAID, 16GiG ECC RAM DDR3 400MHz @1.35V seem rock solid and i don't see the benefit of memory I/O at 10GB/s@666MHz instead of 8GB/s@400MHz)
Even if modern systems or the KGPE-D16 in question were slightly faster i had rarely seen a workstation system under full load which had met the power and efficiency target!
Soon i'll test a Phenom2 X6 95Watt CPU (instead of the typical 125W) which may bring down consumption another 10 or 20 Watts, under constant full load that is. Too i am planning to introduce an option for easy undervolting with coreboot, fixing Northbridge frequency at 1.8GHz maybe with reduced voltage and fixing CPU at 1.25Volts which should be stable with all fam10h CPUs in non-TurboCore performance state so no specific low-power CPUs had to be acquired, and the power management chaos may simply be removed entirely, which was the part that contributed to breaking AM3 fam10h revE support, and it is a pain elsewhere too i might add. s2disk is functional still and system can be completely powered off and restored with it, with encrypted swap space even.
Furthermore such an AM3 system can easily be cooled and remains almost dead silent, simply choose a 5Volt adapter with appropriate 120mm ventilators - it is a bliss, if you had ever enjoyed the noise of a workstation/server nearby.
Sadly fam15h support on AM3 allthough present on branch 4.11 couldn't be confirmed, at least the FX 8120 CPU i had tested on the AM3 board with coreboot didn't POST at all.

Other than that i am maintaining a self-made hardened livedvd distribution booting read-only ISO only targetting these "obsolete" x86-64 systems in conjunction with coreboot which has removed certain "features" in kernel and userspace entirely and it is being stabilized after 3 years of development and testing.
Since fam10h CPUs lack accelerated AES encryption for example the distro of mine ships with a crypto stack inspired by Telco scramblers or MIL-STD-188 this easily yields almost 10GiB/s of randomness and integrated with block layer encrypted disk i/o can be had at 1GiB/s per cpu core max instead, maybe with NILFS2 on top of it to be tested still, without any AES hardware acceleration required anymore. so much for the achievable performance and reliability.

Unrelated but since libpayload was mentioned i'll keep an eye on that one too for sure, i had encounted some minor issues recently which i couldn't track down yet.

**Comment by tlaurion** [comment]
Community attempts to bring the Heads's KGPE-D16 board config to coreboot 4.11 [here](https://github.com/osresearch/heads/pull/795). Any help welcome.

**Comment by tlaurion** [comment]
kgpe-d16 is supported on Heads on coreboot 4.11 with TPM support. That was merged under https://github.com/osresearch/heads/commit/1661e5dcb07a061317581187b15ae553ad8fc3df, thanks to @Tonux599 !

**Comment by tlaurion** [comment]
BMC internal flashing to be tested with #943

**Comment by kuleszdl** [comment]
Thank you for bringing up this issue (I just stumbled over it). I am one of these "complaining" users who has been running this board in production for several years now and figured out a way to live / workaround around the aforementioned issues. Personally, I still find the board unique due to the arguments presented by @tlaurion, and I find its power consumption (at least with the vendor bios) acceptable for a workstation/server (however, not acceptable when running with coreboot and two cpu packages).

Imho, if 3mdeb is willing to take on this work and just needs funding, this is already a good starting point. Yet, collecting 30k is probably too much if you consider the remaining KGPE-D16 owners. But if the 30k would be split into separate goals it might be possible to collect enough money to reach at least some of them.

So, what could be the goals? Probably (some) of these:

- implement necessary features in fam15h to re-upstream it (from the release notes: relocatable ramstage, postcar stage and C_ENVIRONMENT_BOOTBLOCK)
- fix idle power consumption / sleep state issues
- fix memory training issues
- fix related memory training issues on KCMA-D8 as well (which does not work with ECC memory)
- fix booting with microcode updates loaded by coreboot
- fix other bugs?

@miczyg1 Have you considered starting a campaign on CrowdSupply or a similar platform? It's hard to guess how many people would be willing to back such a campaign and of course there is risk to loose the effort invested in the campaign, but imho it's worth a try.

**Comment by Tonux599** [comment]
@kuleszdl If your on the OSWF Slack join #kgpe-d16 where there is discussion on a possible hackathon being organised by 3mdeb. (I can also add you to #kgpe-d16-private where the bulk of the discussion occurred before opting for a public channel)

**Comment by kuleszdl** [comment]
@Tonux599 I don't use slack and usually I prefer discussing such issues more in-depth over public issues (such as here) or on mailing lists. Yet, I am very pleased to hear about the planned hackathon which sounds like a very promising idea!

**Comment by duncancmt** [comment]
I am Duncan Townsend, CTO of [Immunefi](https://immunefi.com/). We love what you're doing. Immunefi would like to pledge your choice of 17,750 [USDC](https://etherscan.io/token/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48) (approx 15,000 EUR) or 10 Ether (approx 22,700 EUR) towards the completion of this work. You can view the pledged funds at [0xD24F38eCe7084F8bAB4BEc23cbcC9a50566Bc38C](https://etherscan.io/address/0xd24f38ece7084f8bab4bec23cbcc9a50566bc38c) or at [kgped16pledge.immunefi.eth](https://etherscan.io/address/kgped16pledge.immunefi.eth). 

Would you like for us to raise awareness of this need throughout the crypto community? We believe this is something that everybody in crypto should be using and want to make sure this happens. Let us know how we can cross-promote and raise awareness.

Please send an email to team@immunefi.com if further coordination is required.

**Comment by pietrushnic** [comment]
Hi @duncancmt,
I'm Piotr Król, Founder and CEO of 3mdeb Embedded Systems Consulting, which initially tried to gather founds through NLNet.

This pledge very generous from your side. Thank you very much.

Rising awareness by respected member of crypto community would definitely will be helpful, we can also use 3mdeb marketing channels to bring enough founds to start this project. With this pledge we think we are very close to deliver something meaningful.

I think other participants here, especially @tlaurion, and members on kgpe-d16 channel of OSFW Slack can also help in promotion.

**Comment by macpijan** [comment]
@duncancmt @tlaurion 

Hello,

Below some notes from our internal meeting:

As far as we understand, the main goals for this project are:
1. RoT in bootblock
2. Platform usage via the BMC
3. Upstream?
4. Maintenance

1. RoT in bootblock
  - the only option we can see is the read-only SPI flash
  - there are two options of achieving this:
    - use a flash chip with OTP feature
    - use hardware mod and WP (write protect) pin of the SPI flash (after
      confirming that it cannot be driven from software in any way)

2. Platform usage via the BMC
  - what exactly is expected from the BMC?
  - we need to refresh the Project Charter in this area
  - https://www.mail-archive.com/coreboot@coreboot.org/msg54222.html
  - https://github.com/u-root/u-bmc/issues/133
  - AST2050 not supported in recent Linux kernel - out-of-tree drivers for
    2.6.29.6 only
   - important bit:
     - there is no SPI flash for BMC by default on the board
     - external SPI flash on extension module is required
     - hardware design: TBD link

3. Upstream or not?

  In the case of the upstream, we need to implement all of the missing features
  that the coreboot 4.14 requires and try to push the board back upstream. When
  not upstreaming, we might implement only the features we need (e.g.
  C_ENVIRONMENT_BOOTBLOCK) on the 4.11 branch.

  - pros
    - in line with our philosophy of upstream development
    - we are forced to include more features (but do all of them lead us to the
      main goals of this project?)
    - better marketing
  - cons
    - potential pushback from the community (upstreaming old board)
    - more effort than improving the state of the 4.11 branch
    - risk that the board might get dropped again after some time if not being
      adjusted to the latest coreboot standards and features

4. Maintenance?
  - Proposal: 6-month release cycle in the next 2-3 years

We would like to set up an external meeting to discuss the above and the remaining
items. The main goal of the meeting would be to confirm the minimum required
scope of the work to make sure that we can reach your goals.

Proposed meeting dates:
1. Wednesday 18th August 15:00 UTC
2. Thursday 19th August 20:00 UTC
3. Monday 23rd August 15:00 UTC
4. Tuesday 24th August 20:00 UTC

Please let me know if any of the proposed dates are fine for you.

**Comment by ln2max** [comment]
@pietrushnic has proposed joining my [flashrom write-protect bounty](https://github.com/flashrom/flashrom/issues/185) to this... which I'm happy to do.

**Comment by macpijan** [comment]
The selected meetin date is: Monday 23rd August 15:00 UTC
The meeting will take place on: https://meet.3mdeb.com/kgpe-d16-refresh

Anyone interested in this project is invited to join.

**Comment by macpijan** [comment]
Slides: 
[kgpe-d16-refresh.pdf](https://github.com/osresearch/heads/files/7032987/kgpe-d16-refresh.pdf)

**Comment by macpijan** [comment]
Let's schedule the 2nd meeting at: Friday 17th September 16:00 UTC

**Comment by macpijan** [comment]
Slides for today's meeting attached. I will post the meeting minutes after the meting as well.

**Comment by macpijan** [comment]
Report from Friday's meeting below.

Slides:
- Duncan: proposal to separate coreboot effort from the BMC effort

Notes from the discussion:

## Participants

There were 9 (? sorry if I have missed anyone - I have not dumped the whole
list) participants. Thank you all for joining this call. The list is below:

1. MDr164 - Marvin, 9elements
2. Tonux - Thomas, worked with kgpe-d16 workstation in heads
3. Piotr Król - 3mdeb
4. Maciej Pijanowski - 3mdeb
5. Michał Żygowski - 3mdeb
6. Krystian Hebel - 3mdeb
7. duncancmt - Duncan Townsend, Immunefi
8. Leah Rowe - Libreboot
9. Anonymous participant

## Discussion

### Decisions

- There is approval for the `critial phases`
  (`coreboot refresh for the KGPE-D16 board ` and `Firmware Root of Trust`)
- The `KCMA-D8` is not a priority for Immunefi - we would rather go with the
  maintenance releases for the `KGPE-D16` if there are funds left
  - we may gather additional funds for the `KCMA-D8` work
  - anyone is invited to contribute with the `KCMA-D8` support - after the
    `KGPE-D16` will be contributed as part of this project, it should be far
    easier to do

### Next steps

- 3mdeb will divide the accepted work into milestones and present the roadmap
  this week - preferably with some monthly deliverables

### Others

- All of you is welcome to join `immunefi.3mdeb` on Keybase
- Libreboot is interesed in incorporating this work and supporting with testing
- FSF is potentially interested with testing support as well
- another reference for this project: https://notabug.org/libreboot/lbmk/issues/48

**Comment by macpijan** [comment]
Hello. The proposed schedule is attached. From our side, we are ready to start with this exciting project!

[kgpe-d16-refresh-schedule.pdf](https://github.com/osresearch/heads/files/7220053/kgpe-d16-refresh-schedule.pdf)

**Comment by zamaudio** [comment]
Hi @pietrushnic, what I meant was: if you are looking for a baseline version to base the ram code off, make sure you choose a commit that was boot tested on actual hw otherwise you may introduce a bug that will be hard to find and may have been already fixed in a different coreboot commit.

**Comment by lss4** [comment]
I was able to boot the latest commit of `4.11_branch` at the time (last year) on KGPE-D16 with a select few RAM combinations. Just that I wasn't able to find a working RAM combination for KCMA-D8.

I think the current state of the `4.11_branch` should still be relevant as there are very few commits directly relevant to our boards, nor did I see anything useful backported to it.

For now I think addressing the memory initialization issues should be the most important at the moment, as it prevents the boards from booting reliably. Other issues (such as USB and system stability) could follow once the board can be reliably booted.

**Comment by pkubaj** [comment]
I use two KGPE-D16's, one is my workstation, another is my server. Both are running coreboot 4.11.

Both are booting reliably and I can't really understand why people say those boards are unstable with coreboot. The initial setup was quite a pain, but after that, they are VERY stable and boot reliably 99% of times (the times where they don't boot, they instantly reboot after loading kernel, but it's quite rare).

**Comment by lss4** [comment]
> I use two KGPE-D16's, one is my workstation, another is my server. Both are running coreboot 4.11.
> 
> Both are booting reliably and I can't really understand why people say those boards are unstable with coreboot. The initial setup was quite a pain, but after that, they are VERY stable and boot reliably 99% of times (the times where they don't boot, they instantly reboot after loading kernel, but it's quite rare).

It depends on the RAM installed, as well as the hardware people installed on the board, as well as what people use coreboot for (Heads or not).

In my experience USB is mostly broken with coreboot. I can't properly boot from USB sticks (for installing a Linux distro), and have to use PS/2 input devices because USB ones don't work once booted into something that has its own USB stack. It also prevented `nvramcui` from running because it couldn't initialize "EHCI periodic frame table". Booting `nvramcui` without any USB devices attached would report different errors that ultimately led me to a `malloc` failure. As I didn't have an OS nor optical drives installed at that time, the USB issues blocked me from testing any further.

While some RAM combinations are near 100% bootable, the Memtest86+ I built into it reported wrong SPD values and detected some unexpected errors then eventually froze with a glitched screen. With less RAM installed (<32GB) it would not detect errors but the test still freezes after a while.

I don't use Heads at the moment, just coreboot in overall. Maybe coreboot+Heads can work reliably once configured properly, but coreboot as it currently is now is not going to be useful as simply getting it to boot on the board is no trivial task for many.

**Comment by pkubaj** [comment]
Ugh, that's really strange. I use USB input devices just fine and can boot from USB just fine (but I don't use Linux, maybe that's why). The coreboot-provided Memtest86+ does report errors right after starting, but AFAIK this is known issue on this board and if you use Memtest86+ booted from another source, it's apparently fine.

I also don't use Heads, since it only supports Linux.

**Comment by lss4** [comment]
> Ugh, that's really strange. I use USB input devices just fine and can boot from USB just fine (but I don't use Linux, maybe that's why). The coreboot-provided Memtest86+ does report errors right after starting, but AFAIK this is known issue on this board and if you use Memtest86+ booted from another source, it's apparently fine.
> 
> I also don't use Heads, since it only supports Linux.

Something else to mention.
1. USB worked while in SeaBIOS so I could use it to select a boot option. It also worked in boot menus like GRUB. However, that's all where it could work. After I booted into something (like MenuetOS), USB input devices stop working completely, so I have to use PS/2 input devices from that point on.
2. I tried booting some USB sticks containing Linux distros. They all failed before reaching the phase to load the kernel.
3. Memtest86+ would freeze regardless of the form: as a coreboot payload, as a floppy image payload, or from USB stick (such as those shipped with Linux distros).
4. I also tried booting some DOS floppies images as payloads (with the help of Mike Banon's patches), and they didn't work either (hangs at some point). Don't know if the issue is caused by memory not being initialized properly, or that SeaBIOS is not behaving in a way some DOS kernels expect (so far no DOS images are working). I'm using a 16MB (128Mbit) flash chip for coreboot by the way.
5. So far I only managed to get MenuetOS (as floppy image payload) booted to a usable state, just without USB keyboard/mouse.

I don't know about your use case though, and I'm not really sure about the status of USB. For me it definitely was problematic.

**Comment by pietrushnic** [comment]
> Hi @pietrushnic, what I meant was: if you are looking for a baseline version to base the ram code off, make sure you choose a commit that was boot tested on actual hw otherwise you may introduce a bug that will be hard to find and may have been already fixed in a different coreboot commit.

Yes, this is fair point. I believe @miczyg1 can say more about his approach to selecting correct code base that will be bring up upstream. There is some limit to number of memory configurations we can test, but we will get to call for beta testers at some point I believe.

**Comment by pietrushnic** [comment]
> I also don't use Heads, since it only supports Linux.

@pkubaj petitiboot is able to kexec FreeBSD so why not Heads :)

**Comment by pkubaj** [comment]
petitboot is able to kexec FreeBSD kernel, but this still has some issues.
1. Skiroot uses kexec-lite from https://github.com/open-power/kexec-lite which is not the kexec most distros use, so there may be some differences,
2. FreeBSD uses its own loader on amd64, but Petitboot can't load loader. It loads just the kernel, which means we can't easily get ZFS on root on powerpc64* systems (unless one builds zfs module in the FreeBSD kernel). It's also not possible to load any other module at boot time (from /boot/loader.conf) or change kernel variables writable at boot-time (one can work it around by adding them to the kernel command line, but I wouldn't call that the FreeBSD way).
Having to load directly the kernel has one big disadvantage - it's not possible to have FDE. /boot needs to be unencrypted, to read kernel. When using FreeBSD loader, it's possible to have the 1st stage loader in the MBR, which then decrypt the HDD and chainload another loader. This is not possible with loading kernel directly.
On POWER systems, there's also the issue that /boot needs to be FAT32, because Skiroot's Linux can't read FreeBSD's UFS or ZFS...

**Comment by pietrushnic** [comment]
@macpijan cc

@pkubaj those are interesting insights, but it seem so far there was no commercial interest in fixing above issues. We are little bit off-topic here, but maybe 3mdeb at some point could fix those issues.

**Comment by pkubaj** [comment]
@pietrushnic
The Talos board I have at home was bought (along with the CPU and some RAM) by FreeBSD Foundation for me to work on FreeBSD Ports on POWER.
I explicitly requested a grant from them and it got approved. It took quite some time to get it, but I THINK it would be possible for you to request a grant to work on FreeBSD on POWER (kernel / OS side). It's another thing whether that gets approved, though. But I'd be happy if there were more OS devs for POWER on FreeBSD, since while on the ports side we're pretty good (Firefox, Libreoffice etc.), we're pretty much behind in terms of OS / kernel (no DRI, no bhyve, no loader and performance is quite behind Linux - even my KGPE-D16 can compile things faster than Talos on FreeBSD).
Anyway, the point was that you MAY be able to get some sponsorship from FreeBSD Foundation regarding POWER.

**Comment by pietrushnic** [comment]
@pkubaj thank you, we will look into sponsorship of FreeBSD Foundation.

**Comment by SergiiDmytruk** [comment]
Request for anyone interested in using KGPE-D16 with write-protection of flash chip: if your chip is not already listed in https://github.com/flashrom/flashrom/issues/185, please post a comment there stating chip's model.

**Comment by macpijan** [comment]
In terms of the October deliverables:
- we plan to deliver Phase 2 as planned
- we plan to deliver Phase 3 except for the `IMM1.6 - KGPE-D16 firmware shall support resource allocator v4` task - we are still working on that and plan to move it to Phase 4 for November instead

We plan the Phase 2 + Phase 3 release for this week.

**Comment by pietrushnic** [comment]
@macpijan do you think it make sense to ask Libreboot community for testing?

**Comment by githubisnonfree** [comment]
hi. leah here. i'm currently in the process of building myself a d16 again, precisely because of your efforts. i will be available for testing. Tashtari in libreboot IRC also has D16 hardware and is quite enthusiastic about your work.

When your work is ready, I wish to integrate it in lbmk, which is libreboot's build system. It can download coreboot, applying any number of custom patches desired. lbmk currently uses standard coreboot 4.11

**Comment by tlaurion** [comment]
Will buy a tpm2 module and can test myself as well with heads.

I will try my best to bring a u-root board encompassing cpu toolset  as well, since I intend to use that board as my local build system and use those 32 cores and 64gb of ram.... And heat some space doing so in the cold weather here.

The kgpe-d16 in current state works already as a q4.1 platform, while some pushing would be needed to have network-server cascade forwarding of external traffic to qubes.

I will reenable that machine that sleeps for way too long.

**Comment by macpijan** [comment]
> do you think it make sense to ask Libreboot community for testing?

It makes perfect sense, but maybe more after a fourth phase (somewhere early December) when we can actually boot something.

> How much would that u-bmc for ast2050 chip additional work cost?

I haven't really used u-bmc, but if it also needs U-Boot and Linux for AST2050 (which I assume it does) then most of the work between OpenBMC and u-bmc would be common. We have presented our view on that in this presentation: https://github.com/osresearch/heads/issues/719#issuecomment-921877456 (starting with slide 13).

I guess you could also try to use the existing kernel, but you would be stuck with the 12 years old kernel running on your BMC then.

**Comment by pietrushnic** [comment]
FYI v0.1.0 was released: https://docs.dasharo.com/variants/asus_kgpe_d16/releases/

Please note you can join [newsletter](https://newsletter.3mdeb.com/subscription/ozes4Jxuo) to get notifications.

Please let us know what we can improve.

**Comment by pkubaj** [comment]
Are there plans to push this work to upstream or will it be maintained as a fork?

**Comment by pietrushnic** [comment]
@pkubaj please take a look at [mileston 7](https://github.com/osresearch/heads/issues/719#issuecomment-926063212). In short we definitely plan to upstream all code that would be accepted by upstream, but let's be honest timeline of that is unknown.

**Comment by Matthew-Bradley** [comment]
A heads up: Over in the mainline they're looking at possibly dropping quite a large number of AMD platforms (fam 14h, 15h, 16h?) if they can't move over to the newer resource allocater (though for 15h there may be some early PoC?). It looks like this work will need to support the new resource allocator in order to be merged.

**Comment by macpijan** [comment]
The allocator v4 was in scope of this project and is being worked already. There are still a couple of more features required in this last deprecation call (like the PARALLEL_MP) which were not part of the initial plan. We will do our best to implement it as part of the upstream effort, though.

**Comment by awokd** [comment]
Is there a separate issue somewhere for PARALLEL_MP/LEGACY_SMP_INIT removal? I've been looking into it a bit as it relates to fam14-16 deprecation.

**Comment by pietrushnic** [comment]
@awokd we all have to be professional in light of what is going on mailing list [1](https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/EEEBFATYHWIPRDXLCUEFNDZ4FYYVA4QM/) [2](https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/LWRE3RQRP7KFTAGLJLUDBX72UA74A5LD/). We tried to present 3mdeb position at [coreboot leadership](https://docs.google.com/document/d/1NRXqXcLBp5pFkHiJbrLdv3Spqh1Hu086HYkKrgKjeDQ/edit#heading=h.682pqlkrvkx3). We will work closely with community to avoid what was proposed. We also consider various backup plans, if you are intrested in discussing those please join [Dasharo Matrix workspace](https://matrix.to/#/#dasharo:matrix.org)

**Comment by Rudd-O** [comment]
> The kgpe-d16 in current state works already as a Qubes 4.1 platform, while some pushing would be needed to have network-server project cascade forwarding of external traffic to specific qubes(@Rudd-O) coming from hidden services down to their actual offered service provided in services qubes, as well as being able to provide some yunohost or equivalent services locally. That's where I was in the past, and where I personally want this to go forward.

I believe Qubes network server is already available for r4.1 and I know I tested it on my throwaway laptop before releasing it on Github:

https://github.com/Rudd-O/qubes-network-server

**Comment by tlaurion** [comment]
@Rudd-O I just saw new commits since last time I checked. Didn't know cascading was implemented, will test and sorry for the noise if it was implemented !

**Comment by Rudd-O** [comment]
Cascading is not explicitly implemented in Qubes 4.1, although the current implementation may just support it.  Please feel free to try!

**Comment by tlaurion** [comment]
> @awokd we all have to be professional in light of what is going on mailing list 1 2.

@pietrushnic your link in 2 points to non-existing content

**Comment by tlaurion** [comment]
Some trails on 4.11 problems :
- gathering from miczyg https://github.com/osresearch/heads/issues/719#issuecomment-636687148
- Paul Menzel: https://github.com/osresearch/heads/issues/719#issuecomment-658652781 "Random boot failures, all of memory not supported, higher energy usage compared to vendor firmware, problems loading latest microcode"
- Recap on current status by myself on 4.11: https://github.com/osresearch/heads/issues/719#issuecomment-660550581, when TPM and coreboot patches were added to Heads by tonux : https://github.com/osresearch/heads/issues/719#issuecomment-743802259
- Ram init issues: https://github.com/osresearch/heads/issues/719#issuecomment-660767252
- Maximum memory support being theoretically 256GB, but tested? https://www.raptorengineering.com/coreboot/kgpe-d16-status.php https://libreboot.org/docs/hardware/kgpe-d16.html#memory-compatibility-with-libreboot

That is what I could grab from skimming both my own memory and traces.
Anyone has other problems still current on 4.11  to report, or other places we might have missed at this point?
@pietrushnic @Th3Fanbus @githubisnonfree ?

**Comment by githubisnonfree** [comment]
neox on irc reported raminit issues on their board; works in coreboot
4.6 but not coreboot 4.11. i'm having that person do a git bisect to
find when coreboot broke, between 4.6 and 4.11, and will report this to
you all when it is found


On Thu, 16 Dec 2021 06:40:48 -0800
tlaurion ***@***.***> wrote:

> Some trails on 4.11 problems :
> - gathering from miczyg
> https://github.com/osresearch/heads/issues/719#issuecomment-636687148
> - Paul Menzel:
> https://github.com/osresearch/heads/issues/719#issuecomment-658652781
> "Random boot failures, all of memory not supported, higher energy
> usage compared to vendor firmware, problems loading latest microcode"
> - Recap on current status by myself on 4.11:
> https://github.com/osresearch/heads/issues/719#issuecomment-660550581,
> when TPM and coreboot patches were added to Heads by tonux :
> https://github.com/osresearch/heads/issues/719#issuecomment-743802259
> - Ram init issues:
> https://github.com/osresearch/heads/issues/719#issuecomment-660767252
> - Maximum memory support being theoretically 256GB, but tested?
> https://www.raptorengineering.com/coreboot/kgpe-d16-status.php
> https://libreboot.org/docs/hardware/kgpe-d16.html#memory-compatibility-with-libreboot
> 
> That is what I could grab from skimming both my own memory and traces.
> Anyone has other problems still current on 4.11  to report, or other
> places we might have missed at this point? @pietrushnic @Th3Fanbus
> @githubisnonfree ?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/osresearch/heads/issues/719#issuecomment-995879031
> You are receiving this because you were mentioned.
> 
> Message ID: ***@***.***>


-- 
Leah Rowe ***@***.***>
Company Director & Libreboot developer

Do you know you have rights?
The right to privacy, free speech, the right to read
and the right to learn.

Defend freedom. Use free (free as in freedom) software.
Spread freedom. Tell everyone you know about it!
https://www.gnu.org/philosophy/free-sw.html

Minifree Ltd, trading as Ministry of Freedom.
Registered in England, registration No. 9361826
VAT Registration No. GB202190462
Minifree Ltd, 19 Hilton Road, Canvey Island
Essex SS8 9QA, United Kingdom
United Kingdom

**Comment by githubisnonfree** [comment]
neox on irc reported raminit issues on their board; works in coreboot
4.6 but not coreboot 4.11. i'm having that person do a git bisect to
find when coreboot broke, between 4.6 and 4.11, and will report this to
you all when it is found

**Comment by githubisnonfree** [comment]
by the way i'm not sure if this is addressed but:

is the re-upstreaming work also being done on kcma-d8? the d8/d16 codebases are very similar in coreboot

**Comment by macpijan** [comment]
> being done on kcma-d8? the d8/d16 codebases are very similar in coreboot, and i think it would be worthwhile.

It is not part of this effort. But once KGPE-D16 code is refreshed, it would be much easier.
It was evaluated, but did not fit into the scope. See slide 23 in the presentation: https://github.com/osresearch/heads/issues/719#issuecomment-921877456

**Comment by tlaurion** [comment]
15h agesa fork has issues with tpm under linux as of today.

**Comment by tlaurion** [comment]
https://github.com/linuxboot/heads/issues/134

**Comment by tlaurion** [comment]
https://github.com/linuxboot/heads/issues/134
